lib/teleterm/vnet/service.go

```diff
@@ -274,3 +330,120 @@ func (p *appProvider) newTeleportClient(ctx context.Context, profileName, leafCl
 	}
 	return tc, nil
 }
+
+type usageReporter struct {
+	cfg UsageReporterConfig
+	// reportedApps contains a set of URIs for apps which usage has been already reported.
+	// App gateways (local proxies) in Connect report a single event per gateway created per app. VNet
+	// needs to replicate this behavior, hence why it keeps track of reported apps to report only one
+	// event per app per VNet's lifespan.
+	reportedApps map[string]struct{}
+	// mu protects access to reportedApps.
+	mu sync.Mutex
+}
+
+type clientCache interface {
+	GetCachedClient(context.Context, uri.ResourceURI) (*client.ClusterClient, error)
+}
+
+type eventConsumer interface {
+	ReportUsageEvent(*teletermv1.ReportUsageEventRequest) error
+}
+
+type UsageReporterConfig struct {
+	ClientStore   *client.Store
+	ClientCache   clientCache
+	EventConsumer eventConsumer
+	// clusterIDCache stores cluster ID that needs to be included with each usage event. It's updated
+	// outside of usageReporter – the middleware merely reads data from it. If the cache does not
+	// contain the given cluster ID, usageReporter drops the event.
+	ClusterIDCache *clusteridcache.Cache
+	InstallationID string
+}
+
+func (c *UsageReporterConfig) CheckAndSetDefaults() error {
+	if c.ClientStore == nil {
+		return trace.BadParameter("missing ClientStore")
+	}
+
+	if c.ClientCache == nil {
+		return trace.BadParameter("missing ClientCache")
+	}
+
+	if c.EventConsumer == nil {
+		return trace.BadParameter("missing EventConsumer")
+	}
+
+	if c.ClusterIDCache == nil {
+		return trace.BadParameter("missing ClusterIDCache")
+	}
+
+	if c.InstallationID == "" {
+		return trace.BadParameter("missing InstallationID")
+	}
+
+	return nil
+}
+
+func NewUsageReporter(cfg UsageReporterConfig) (*usageReporter, error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &usageReporter{
+		cfg:          cfg,
+		reportedApps: make(map[string]struct{}),
+	}, nil
+}
+
+func (r *usageReporter) ReportApp(ctx context.Context, appURI uri.ResourceURI) error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	if _, hasAppBeenReported := r.reportedApps[appURI.String()]; hasAppBeenReported {
+		log.DebugContext(ctx, "App was already reported", "app", appURI.String())
```

Comment:
this might be pretty spammy if printed on every tcp connection, but I guess we do already have a bunch of other debug logs for every tcp connection

Reply:
I'll just remove it, it's not really needed. If it was already reported, then you'll see an appropriate log emitted before.